### DISCUSSION TITLE
Question on TileLayer

### DISCUSSION THREAD

**Discussion by mskimN** [visgl/deck.gl#5233] [Q&A]
Hello!

I have a question on the bounds of TileLayer. 

I am fetching cropped tiles from a tile server and provide them to TileLayer's "data".

It works well and it shows my raster file well. However, when my viewpoint moves out of raster files' bounds, 

it shows following error. 

![image](https://user-images.githubusercontent.com/70316208/101441811-45c1c280-395d-11eb-8ebe-67f12eafd333.png)

Would there be any solution to stop tile request when my viewpoint is out of bound?

![image](https://user-images.githubusercontent.com/70316208/101444110-211c1980-3962-11eb-9099-b5e46c45cec4.png)

```
  useEffect(() => {
    fetch(endpoint + raster_path)
      .then((res) => {
        if (res.ok) return res.json();
        // throw new Error("Network response was not ok.");
      })
      .then((data) => {
        setTiles(data);
      });
  }, [raster_path]);

  const tileLayer = new TileLayer({
    data: [tiles.tiles + color],
    maxRequests: 6,

    pickable: true,
    onViewportLoad: null,
    autoHighlight: false,
    highlightColor: [60, 60, 60, 40],
    minZoom: 0,
    maxZoom: 19,
    tileSize: 512 / devicePixelRatio,
    extent: [tiles.bounds],
    maxCacheSize: 30,

    renderSubLayers: (props) => {
      const {
        bbox: { west, south, east, north },
      } = props.tile;
      console.log(props);

      return [
        new BitmapLayer(props, {
          data: null,
          image: props.data,
          bounds: [west, south, east, north],
        }),
      ];
    },
  });

  const layers = [
    tileLayer,
  ];

  return (
    <div>
      <DeckGL
        initialViewState={viewport}
        views={new MapView({ repeat: true })}
        controller={true}
        layers={[layers]}
        onViewportChange={handleViewportChange}
      >
        <ReactMapGL
          {...viewport}
          width="100%"
          height="100%"
          onViewportChange={handleViewportChange}
          mapboxApiAccessToken={process.env.REACT_APP_MAPBOX_TOKEN}
          mapStyle="mapbox://styles/mapbox/streets-v9"
        ></ReactMapGL>
      </DeckGL>
    </div>
  );
}
```

**Comment by Pessimistress** [comment]
You need to implement your own [getTileData](https://deck.gl/docs/api-reference/geo-layers/tile-layer#gettiledata):

```js
import {load} from '@loaders.gl/core';

const layer = new TileLayer({
  // ...
  getTileData: tile => {
    if (tileWithinBounds(tile.x, tile.y, tile.z) {
      return load(tile.url);
    }
    return null;
  }
});
```

FYI don't make both DeckGL and ReactMapGL handle user interaction. See https://github.com/visgl/deck.gl/blob/master/examples/get-started/react/mapbox/app.js for the correct usage pattern.

**Reply by mskimN** [reply]
Thank you for your reply!
Is StaticMap more efficient than ReactMapGL?

**Reply by Pessimistress** [reply]
ReactMapGL (the default export of react-map-gl) implements its own event listeners and map control. StaticMap is the sub component that excludes the interaction logic. When you are using it as a child of DeckGL, you only need the parent to handle user interaction. In fact, if both the deck canvas and the map try to respond to user input, you may get some weird effect as they both attempt to update the viewport.

**Reply by geniuskim05** [reply]
@Pessimistress Is tileWithinBounds function?

**Comment by ojack** [comment]
I am trying to do this using the MVTLayer class, as I have made custom vector tiles that only have a specific bounds. However, it seems that it isn't possible to overwrite the 'getTileData' function in the MVTLayer class, only in the TileLayer class. How would I load vector tiles that are only restricted to a certain bounds?

**Comment by Pessimistress** [comment]
As of 8.5.8 this can be achieved by setting [extent](https://deck.gl/docs/api-reference/geo-layers/tile-layer#extent)

**Reply by ojack** [reply]
Should extent be in latitude / longitude coordinates or something else? It doesn't have an effect for me when I use it. I am using the latest bundled deckgl version accessed via:  `<script src="https://unpkg.com/deck.gl@latest/dist.min.js"></script>`

Thank you!

**Reply by Pessimistress** [reply]
Example: https://codepen.io/Pessimistress/pen/VwzQPyP?editors=001
